### PR TITLE
Update logging

### PR DESCRIPTION
**Pull request by andersbogsnes**
- [x] closes issue #126 #166 
- [x] Tests added
- [x] Passes flake8
- [ ] Updated CHANGELOG
- [ ] Updated README.md

Implements dumping and loading model class to YAML. Also implements multimetric scoring

Refactors internals to better support this.

### CODE REVIEW

File: src/ml_tooling/baseclass.py [hunk 1]
```diff
@@ -139,20 +163,47 @@ def save_estimator(self, path: str) -> pathlib.Path:
...
+            log.save_log(self.config.RUN_DIR)
 
         logger.info(f"Saved estimator to {estimator_file}")
 
         return estimator_file
 
+    def dump(self):
+        if self.is_pipeline:
+            return [
+                {
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: How can you tell whici parts of the pipe connects where?

Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: `pipeline.steps` is a list of transformations, so the order is fixed

File: src/ml_tooling/metrics/metric.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    metric: str = attr.ib()
+    score: Union[float, int] = attr.ib(default=None)
+    cross_val_scores: np.ndarray = attr.ib(default=None)
+
+    def score_metric(self, estimator, x, y):
+        scoring_func = get_scorer(self.metric)
+        self.score = scoring_func(estimator, x, y)
+
+    @property
+    def std(self):
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/metrics/metric.py]: I know Sci-Kit learn does this to but you may grossly underestimate variance, http://papers.nips.cc/paper/2468-no-unbiased-estimator-of-the-variance-of-k-fold-cross-validation.pdf . An idea is avalible at https://stats.stackexchange.com/questions/31190/variance-estimates-in-k-fold-cross-validation

Comment (andersbogsnes) [reply src/ml_tooling/metrics/metric.py]: Agreed - we have an issue #158 on it already, just want to do a 1-1 in this pull requet

File: src/ml_tooling/baseclass.py (hunk 1 above)
Comment (thomasfrederikhoeck) [reply src/ml_tooling/baseclass.py]: We should test that this works with multiple pipliens and unions.

File: src/ml_tooling/utils.py [hunk 3]
```diff
@@ -175,3 +177,44 @@ def _validate_estimator(estimator):
...
     raise MLToolingError(f"Expected a Pipeline or Estimator - got {type(estimator)}")
+
+
+def is_pipeline(estimator: Estimator):
+    if type(estimator).__name__ == "Pipeline":
+        return True
+    return False
+
+
+def setup_pipeline_step(
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/utils.py]: Should have a docsting. What is this used for?

File: src/ml_tooling/baseclass.py [hunk 4]
```diff
@@ -139,20 +161,47 @@ def save_estimator(self, path: str) -> pathlib.Path:
...
+        return [
+            {
+                "module": self.estimator.__class__.__module__,
+                "classname": self.estimator.__class__.__name__,
+                "params": self.estimator.get_params(),
+            }
+        ]
+
+    @classmethod
+    def load(cls, log_file: pathlib.Path):
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: Aren't we making two load functions now? Maybe we should have .load_from_yaml and .load_from_storage

File: src/ml_tooling/baseclass.py [hunk 5]
```diff
@@ -139,20 +161,47 @@ def save_estimator(self, path: str) -> pathlib.Path:
...
-                estimator_path=str(estimator_file),
-            )
+            log = Log.from_result(self.result, estimator_path=estimator_file)
+            log.save_log(self.config.RUN_DIR)
 
         logger.info(f"Saved estimator to {estimator_file}")
 
         return estimator_file
 
+    def dump(self):
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: When we use dump I think one would expect it to be a pickle file. Maybe we should use another name.

File: src/ml_tooling/baseclass.py (hunk 4 above)
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Had changed it to `from_dict`, but that is incorrect - changed it to `from_yaml`

Comment (thomasfrederikhoeck) [reply src/ml_tooling/baseclass.py]: As they both load the estimator - one trained and one untrained. Alternativly we should have a .load and a .specify

File: src/ml_tooling/baseclass.py (hunk 5 above)
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Is now `to_dict`

File: src/ml_tooling/baseclass.py (hunk 1 above)
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Added testcases for Unions and changed code to recursively look in params for Pipelines

File: src/ml_tooling/metrics/metric.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        >>> metric.cross_val_scores
+        array([1., 1.])
+        >>> metric.std
+        0.0
+
+    .. _CV: https://scikit-learn.org/stable/modules/classes.html#module-sklearn.model_selection
+    .. _score: https://scikit-learn.org/stable/modules/model_evaluation.html#common-cases-predefined-values # noqa
+    """
+
+    metric: str = attr.ib()
```
[the remark this answers is not in the document]
Comment (thomasfrederikhoeck) [reply src/ml_tooling/metrics/metric.py]: Or is it the metric name?

File: src/ml_tooling/search/gridsearch.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import joblib
+
+
+def _fit_gridpoint(estimator, params, train_x, train_y):
+    estimator.set_params(**params)
+    estimator.fit(train_x, train_y)
+    return estimator
+
+
+def gridsearch(estimator, train_x, train_y, params, n_jobs):
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/search/gridsearch.py]: Firstly it should not be called `gridsearch` as it doesn't do the gridsearch but only generates the estimators so should probabily be name  `make_estimators_for_gridsearch`. Secondly, why do we train it, when it is trained again on the folds when the gridsearch is done?

File: src/ml_tooling/metrics/metric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andersbogsnes) [reply src/ml_tooling/metrics/metric.py]: It's the metric name - we can change it to that if that's better?

File: src/ml_tooling/search/gridsearch.py (hunk 7 above)
Comment (andersbogsnes) [reply src/ml_tooling/search/gridsearch.py]: Good point, was originally done for non-CV metric, but of course cross_val_score already scores the model. I assume we would never allow for a non-CV gridsearch?

Comment (andersbogsnes) [reply src/ml_tooling/search/gridsearch.py]: Should be fixed now

File: src/ml_tooling/metrics/metric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (thomasfrederikhoeck) [reply src/ml_tooling/metrics/metric.py]: yeah let's change it to `metric_name`.

File: src/ml_tooling/baseclass.py [hunk 8]
```diff
@@ -94,7 +117,7 @@ def load_estimator(cls, path: str) -> "Model":
         logger.info(f"Loaded {instance.estimator_name}")
         return instance
 
-    def save_estimator(self, path: str) -> pathlib.Path:
+    def save_estimator(self, path: Union[str, pathlib.Path]) -> pathlib.Path:
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: This is not reflected in the docstring.

File: src/ml_tooling/baseclass.py [hunk 9]
```diff
@@ -56,13 +51,41 @@ def is_regressor(self):
...
+
+    @default_metric.setter
+    def default_metric(self, metric):
+        if self.is_classifier:
+            self.config.CLASSIFIER_METRIC = metric
+        else:
+            self.config.REGRESSION_METRIC = metric
 
     @classmethod
     def load_estimator(cls, path: str) -> "Model":
```
[the remark this answers is not in the document]
Comment (thomasfrederikhoeck) [reply src/ml_tooling/baseclass.py]: Especially when it says in the docstring ;-)

File: src/ml_tooling/baseclass.py (hunk 8 above)
Comment (thomasfrederikhoeck) [reply src/ml_tooling/baseclass.py]: What happens if file already exists? Maybe we should raise?

**Comment by andersbogsnes** [comment]
Maybe just name?


________________________________

File: src/ml_tooling/baseclass.py [hunk 10]
```diff
@@ -208,40 +250,12 @@ def make_prediction(
...
-        else:
-            self.config.REGRESSION_METRIC = metric
-
     @classmethod
     def test_estimators(
         cls,
         data: Dataset,
         estimators: Sequence,
-        metric: Optional[str] = None,
+        metrics: Union[str, List[str]] = "default",
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: Does this limit us to only using named sklearn metrics?

File: src/ml_tooling/metrics/metric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andersbogsnes) [reply src/ml_tooling/metrics/metric.py]: how about `name`?

File: src/ml_tooling/baseclass.py [hunk 11]
```diff
@@ -277,20 +291,22 @@ def test_estimators(
...
-                result.log_estimator(log_dir)
 
-        results.sort(reverse=True)
-        best_estimator = results[0].estimator
+        results = ResultGroup(results).sort()
+
+        if log_dir:
+            results.log_estimator(pathlib.Path(log_dir))
+
+        best_estimator = results[0].model
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: Should we retrain the best estimator? Or at least give the option?

File: src/ml_tooling/metrics/metric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (thomasfrederikhoeck) [reply src/ml_tooling/metrics/metric.py]: deal 🤝

File: src/ml_tooling/baseclass.py (hunk 8 above)
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Not touching the save_estimator and load_estimator yet, as that is handled by #292

File: src/ml_tooling/baseclass.py (hunk 10 above)
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: At the moment, yes. This was also the behaviour previously, but we can create an issue to modify this in the future. Would probably be handled by a `.from_callable` method on metric

File: src/ml_tooling/baseclass.py (hunk 11 above)
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: I can add that as a keyword flag

Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Done

File: src/ml_tooling/baseclass.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Also waiting for #292
